lisa/tools/fio.py

```diff
@@ -0,0 +1,142 @@
+                "gawk",
+                "mdadm",
+                "wget",
+                "sysstat",
+                "blktrace",
+                "bc",
+                "libaio-dev",
+                "zlib1g-dev",
+            ]
+        elif isinstance(self.node.os, Suse):
+            package_list = ["wget", "mdadm", "blktrace", "libaio1", "sysstat", "bc"]
+        else:
+            raise LisaException(
+                f"tool {self.command} can't be installed in distro {self.node.os.name}."
+            )
+        for package in list(package_list):
+            if posix_os.is_package_in_repo(package):
+                posix_os.install_packages(package)
+
+    def _install_from_src(self) -> bool:
+        self._install_dep_packages()
+        tool_path = self.get_tool_path()
+        self.node.shell.mkdir(tool_path, exist_ok=True)
+        git = self.node.tools[Git]
+        git.clone(self.fio_repo, tool_path)
+        code_path = tool_path.joinpath("fio")
+        git.checkout(
+            ref="refs/heads/master", cwd=code_path, checkout_branch=self.branch
+        )
+        from .make import Make
+
+        make = self.node.tools[Make]
+        make.make_install(cwd=code_path)
+        self.node.execute(
+            "ln -s /usr/local/bin/fio /usr/bin/fio", sudo=True, cwd=code_path
+        ).assert_exit_code()
+        return self._check_exists()
+
+    def install(self) -> bool:
+        posix_os: Posix = cast(Posix, self.node.os)
+        posix_os.install_packages("fio")
+        if not self._check_exists():
+            self._install_from_src()
+        return self._check_exists()
+
+    def launch(
+        self,
+        name: str,
+        filename: str,
+        mode: str,
+        iodepth: int,
+        numjob: int,
+        time: int = 60,
+        block_size: str = "4K",
+        direct: bool = True,
+        gtod_reduce: bool = True,
+        ioengine: str = "libaio",
+        group_reporting: bool = True,
+    ) -> FIOResult:
+        cmd = (
+            f"--ioengine={ioengine} --bs={block_size} --filename={filename} "
+            f"--readwrite={mode} --runtime={time} --iodepth={iodepth} "
+            f"--numjob={numjob} --name={name}"
+        )
+        if direct:
+            cmd += " --direct=1"
+        if gtod_reduce:
+            cmd += " --gtod_reduce=1"
+        if group_reporting:
+            cmd += " --group_reporting"
+        result = self.run(
+            cmd,
+            force_run=True,
+            sudo=True,
+            expected_exit_code=0,
+            expected_exit_code_failure_message=f"fail to run {cmd}",
+        )
+        matched_results = self._result_pattern.match(result.stdout)
+        assert matched_results
+        iops = matched_results.group("avg_iops")
+        latency = matched_results.group("latency")
+        fio_result = FIOResult()
+        fio_result.iops = Decimal(iops)
+        fio_result.latency = Decimal(latency)
+        fio_result.qdepth = iodepth * numjob
```

Comment:
Is the qdepth a common concept? I can see only iodepth in fio command. How about we save both iodepth and cpu count?

Reply:
add iodepth, core_count, disk_count